mlir/test/Dialect/MIOpen/lowering_top_level.mlir

```diff
@@ -28,6 +29,28 @@ func @miopen_conv2d(%filter : memref<128x8x3x3xf32>, %input : memref<128x8x32x32
 // CHECK-NEXT:  miopen.transform(%arg2)
 // CHECK-NEXT:  miopen.gridwise_gemm
 
+func @miopen_conv2d_f16(%filter : memref<128x8x3x3xf16>, %input : memref<128x8x32x32xf16>, %output : memref<128x128x30x30xf16>) {
+  miopen.conv2d(%filter, %input, %output) {
+    arch = "gfx906",
+    num_cu = 64,
+    filter_layout = ["k", "c", "y", "x"],
+    input_layout = ["ni", "ci", "hi", "wi"],
+    output_layout = ["no", "ko", "ho", "wo"],
+    dilations = [1, 1],
+    strides = [1, 1],
+    padding = [0, 0]
+  } : memref<128x8x3x3xf16>, memref<128x8x32x32xf16>, memref<128x128x30x30xf16>
+  return
+}
+// CHECK-LABEL: func {{@miopen_conv2d.*%arg0.*%arg1.*%arg2}}
+// CHECK-NOT:   miopen.conv2d
+// CHECK-NEXT:  miopen.transform(%arg0)
+// CHECK-NEXT:  miopen.transform(%arg1)
```

Comment:
More details for transform?

Reply:
I think this granularity is good. Is there any specific fp16 related lowering that need to check-out here?

Reply:
> as that Op may get optimized away under some cases

At what circumstances does it get optimized away? I do see `FPTruncOp` populated when we do the `threadwiseCopyV2`. However, that is not in this lowering step. This unit test handle to gridwise_gemm lowering only. If this get you confused could you rename those that test this to include `step1` in it. Just so that we know it is `-miopen-lowering-step1`

Reply:
It happens at step4. It is tested in lowering_threadwise_copy_v2_f16.mlir. The trunc Op gets optimized away when the source and destination data types are the same. I guess the truncation is necessary when the input and output tensors have different date types. But so far we haven't seen that case and I don't know if different data types would be supported. Again I'm not 100% sure when the source and destination types might be different.

Reply:
Lowering_top_level.mlir just ensures fp16 is handled at the top level.

Reply:
> It happens at step4. It is tested in lowering_threadwise_copy_v2_f16.mlir.

Make sense. You just confirmed that we don't need additional test case in this unit test.

> But so far we haven't seen that case and I don't know if different data types would be supported. 

https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/021b2dd9c7a22ea563fca19bf281d42390d8eebb/mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h#L4365-L4382

If we are talking about the same thing then it looks to me this truncation always happen for fp16 result type (before the store). Could you add this test case to step4 lowering?

Reply:
Thanks, appreciate helping to check this.